### PR TITLE
Support casting from String to Decimal

### LINKED ISSUE #3280: Support casting from String to Decimal
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`cast` kernel doesn't support casting from String to Decimal now. We should support it.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #3280.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -7104,4 +7300,256 @@ mod tests {
...
+            Decimal128Type::format_decimal(
+                parse_string_to_decimal_native::<Decimal128Type>("12345", 2).unwrap(),
+                38,
+                2
+            ),
+            "12345.00"
+        );
+        assert_eq!(
+            Decimal128Type::format_decimal(
+                parse_string_to_decimal_native::<Decimal128Type>("0.12345", 2).unwrap(),
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7324]: I recommend test coverage for the following cases (apologies if they are already covered but I didn't see them):

1. trim, eg, `". 0.123"`
2. Multiple leading zeros: `"000.123"`
3. Multiple trailing zeros: `"12.234000"`
4. empty string `""`
4. whitespace string `"   "`
5. Invalid `"4.4.5"`

File: arrow-cast/src/cast.rs [hunk 2]
```diff
@@ -7104,4 +7300,256 @@ mod tests {
...
+        ]);
+        let array = Arc::new(str_array) as ArrayRef;
+
+        test_cast_string_to_decimal(array);
+    }
+
+    #[test]
+    fn test_cast_large_utf8_to_decimal() {
+        let str_array = LargeStringArray::from(vec![
+            "123.45", "1.2345", "0.12345", "0.1267", "1.263", "12345.0", "12345",
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7442]: I recommend a test for an input array that has Nulls in it as well

File: arrow-cast/src/cast.rs [hunk 3]
```diff
@@ -7104,4 +7300,256 @@ mod tests {
...
+            decimal_arr.value_as_string(3)
+        );
+        assert_eq!(
+            "100000000000000000000000000000000000.00",
+            decimal_arr.value_as_string(4)
+        );
+    }
+
+    #[test]
+    fn test_cast_utf8_to_decimal128_overflow() {
```
Comment (alamb) [inline arrow-cast/src/cast.rs:7468]: 👍

**Review by alamb — APPROVED** [review]
Thank you @viirya  -- I reviewed the code and tests and they looked good to me 👍 

I had a few suggested additional test cases but otherwise looks great 🏅

File: arrow-cast/src/cast.rs (hunk 1 above)
Comment (viirya) [reply arrow-cast/src/cast.rs:7324]: Thanks. Added more test coverage including these cases and Nulls.

File: arrow-cast/src/cast.rs [hunk 4]
```diff
@@ -2801,6 +2827,176 @@ fn cast_utf8_to_boolean(
...
+        let div =
+            i256::from_i128(10_i128).pow_checked((decimals.len() - scale) as u32)?;
+
+        let half = div.div_wrapping(i256::from_i128(2));
+        let half_neg = half.neg_wrapping();
+
+        let d = decimal_number.div_wrapping(div);
+        let r = decimal_number.mod_wrapping(div);
+
+        // Round result
```
Comment (liukun4515) [inline arrow-cast/src/cast.rs:2869]: 👍

**Pull request merged**